### ISSUE TITLE
tcl command getaccount: returning empty string

### ISSUE THREAD

**Issue by jobe1986** [eggheads/eggdrop#1276]
Ok, so this bug is in part the result of an IRC network lacking WHOX whilst having account-notify and extended-join capabilities.

The main issue here is that getaccount when supplied with a nick and without a channel returns the stored account name for the first channel the specified nick is a member on, without checking other channel memberships to see if the use has a stored account on any of those.

Without WHOX but with extended-join it is possible for a user to have a stored account name in its membership for one channel and not the others. As a result, rather then finding the correctly stored account name, getaccount returns the first result it finds which can be "".

Ideally upon receiving a JOIN with extended-join enabled, the account name for the user would be updated for all other membership records for that user too, the same way account-notify handling does for the ACCOUNT message. Alternatively getaccount without a channel specified should at least check the account stored for all the channels the user is a member of to ensure they are all "" before returning "".

I was able to demonstrate this behaviour using the attached [allaccounts.txt](https://github.com/eggheads/eggdrop/files/8310713/allaccounts.txt) tcl script and getting this output on the party line:
```
[06:01:17] <Jobe> .allaccounts Jobe
[06:01:17] <Enopa> Jobe has account: ""
[06:01:17] <Enopa> Jobe on #unrealircd has account: ""
[06:01:17] <Enopa> Jobe on #linux has account: ""
[06:01:17] <Enopa> Jobe on #magirc has account: ""
[06:01:17] <Enopa> Jobe on #anope has account: "Jobe"
```

**Comment by vanosg** [comment]
Thanks for taking the time to write up such a thorough issue, and using Eggdrop!

From the docs ((https://docs.eggheads.org/mainDocs/tcl-commands.html#getaccount-nickname-channel)

```
getaccount <nickname> [channel]

Returns: the services account name of the nickname if they are logged in, “” otherwise, and an error if the account-notify or
extended-join capabilities are not enabled. WARNING: this account list may not be accurate depending on the server and 
configuration. This command will only work if a server supports (and Eggdrop has enabled) the account-notify and extended-join 
capabilities, and the server understands WHOX requests (also known as raw 354 responses).
```

The last sentence documents this limitation. I don't know that we have a good option for handling this other than notifying developers about this limitation and hoping they extend the requirement to people using their scripts, or including a check whether to run or not. I'll leave this open for others to comment with ideas for a bit, then close if no new ideas are presented.

**Comment by vanosg** [comment]
Sorry, I blurred over your proposed suggestion- there may be a reason why it is done /has to be done the way it is now, we'll take a look. Thanks again

**Comment by thommey** [comment]
The fundamental underlying issue is that we have "channel users" and no concept of a "server user".
So right now, Testuser1 on #chan1 is completely unrelated to Testuser1 on #chan2.
Using a few of our commands (onchan, getaccount, ...) allows to fetch a non-deterministic first record, mostly as a convenience to not have to specify a channel, but it is a flawed concept if the information across all those "channel users" can be inconsistent.

Your proposal assumes that a non-empty account name is guaranteed to be the highest value of information about their account but that is not true in all cases or IRCds.
On Libera, for example, users can log out of their account, making "" the most recent and up-to-date information, but your approach would try to find their "old" account" which might have undesired effects.

We could improve consistency by propagating new information about a "channel user" to all their entries in all channels, so it does not matter which one our functions choose, it is always the newest information.
We already do this for things that we receive globally, like nick changes, away status changes, etc.
So I could see it being a sane concept for accounts, too.

In the future a better concept might be to have "server users" as real IRCds handle it with only one place where we store their account.

**Comment by vanosg** [comment]
I'm debating in my head if there is a problem doing this if extended-join is enabled but not WHOX (with "this" being updating other channel records on an extended-join). It certainly gets to be nasty code logic trying to cover every corner case with these non-coordinated ircv3 capabilities, and it's a shame that this gets pushed to the clients to try and muddle their way through in terms of support... but I digress...

Is there something we would break in the expected-case where all three capabilities are present if we were to search for 'nick' on all existing channel records when an extended-join is seen? I can't think of one, and I can't think of a reason why extended-join would have less-current information than Eggdrop when all capabilities are enabled. 

In short, my priorities are- not break something when all capabilities are enabled, then make scenarios where not all 3 are enabled 'less worse' without further breaking other scenarios of half-workingness.

**Comment by jobe1986** [comment]
Theoretically the account field of extended-join should ALWAYS be the most up to date value for a users account name as this is taken from the IRCd's user/client record every time. So if an extended-join's value is to be taken as up to date, applying that value to all user records for the matching user would not be incorrect behaviour, and should not break anything.

**Comment by thommey** [comment]
For "global events" like nick changes, ACCOUNT messages, AWAY notifications, CHGHOST, etc. we propagate the information for each channel record of a nickname, so we already rely on the assumption that we can identify a network-wide user by their nickname.

Applying that behavior for events that are per-channel but are of interest in other channels as well (because the user is also there, identified by nickname) seems fine to me.
Ideally using the opportunity to take a look at what other events could be in that category, other than extjoin with account information.

**Comment by ZarTek-Creole** [comment]
I allowed myself to create a small TCL code to circumvent the problem for the moment. I hook the issue a bit by offering you the command [getaccountAllChan \<nickname\> \[channel\]](https://github.com/ZarTek-Creole/tcl_getaccountAllChan)

If [channel] is provided, it will act as the eggdrop's getaccount

If only \<nickname\> is provided, the eggdrop will search all the channels they have in common and return a list

Thank's to [eggheads team](https://github.com/eggheads) and [jobe1986](https://github.com/jobe1986)
